ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt

```diff
@@ -281,6 +294,58 @@ class RuleEvalTest {
         assertFailsWith<IllegalArgumentException> { eval("""exp("1")""") }
     }
 
+    @Test
+    fun ip() {
+        assertEquals(167772160f, eval("""ip("10.0.0.0")"""))
+        assertEquals(2560f, eval("""ip("0.0.10.0")"""))
+        assertEquals(2165248f, eval("""ip("0.33.10.0")"""))
+        assertEquals(3232238083f, eval("""ip("192.168.10.3")"""))
+        assertEquals(4294967041f, eval("""ip("255.255.255.1")"""))
+
+        // Test IP composition
+        assertEquals(1f, eval("""gt(ip("192.168.1.0"), ip("192.168.0.255"))"""))
+        assertEquals(1f, eval("""gte(ip("192.168.0.255"), ip("192.168.0.255"))"""))
+        assertEquals(1f, eval("""lt(ip("255.0.255.255"), ip("255.1.0.0"))"""))
+        assertEquals(1f, eval("""lt(ip("9.255.255.255"), ip("10.0.0.0"))"""))
+
+        assertFailsWith<IllegalArgumentException> { eval("""ip("10.0.0")""") }
+        assertFailsWith<IllegalArgumentException> { eval("""ip("10.0.0.0.0")""") }
+        assertFailsWith<IllegalArgumentException> { eval("""ip("10.0.0.-1")""") }
+        assertFailsWith<IllegalArgumentException> { eval("""ip("10.0.0.256")""") }
+        assertFailsWith<IllegalArgumentException> { eval("""ip("10.0.0.a")""") }
+    }
+
+    @Test
+    fun cidr() {
+        // testing reference from https://mxtoolbox.com/subnetcalculator.aspx
+
+        // this syntax is not supported, will always return 0
+        assertEquals(0f, eval("""contains([cidr("254.200.222.210/23")], ip("254.200.222.25"))"""))
+
+        assertEquals(0f, eval("""contains(cidr("254.200.222.210/23"), ip("254.200.224.0"))"""))
+        assertEquals(0f, eval("""contains(cidr("254.200.222.210/23"), ip("254.200.221.255"))"""))
+        assertEquals(0f, eval("""contains(cidr("254.200.222.210/23"), ip("254.200.200.10"))"""))
+        assertEquals(0f, eval("""contains(cidr("192.167.233.10/28"), ip("192.167.233.16"))"""))
+        assertEquals(0f, eval("""contains(cidr("192.167.233.10/28"), ip("192.167.232.255"))"""))
+        assertEquals(0f, eval("""contains(cidr("192.167.233.10/28"), ip("192.167.255.16"))"""))
+        assertEquals(0f, eval("""contains(cidr("192.167.233.11/32"), ip("192.167.233.12"))"""))
+
+        assertEquals(1f, eval("""contains(cidr("254.200.222.210/23"), ip("254.200.223.255"))"""))
+        assertEquals(1f, eval("""contains(cidr("254.200.222.210/23"), ip("254.200.222.0"))"""))
+        assertEquals(1f, eval("""contains(cidr("254.200.222.210/23"), ip("254.200.222.188"))"""))
+        assertEquals(1f, eval("""contains(cidr("192.167.233.10/28"), ip("192.167.233.15"))"""))
+        assertEquals(1f, eval("""contains(cidr("192.167.233.10/28"), ip("192.167.233.0"))"""))
+        assertEquals(1f, eval("""contains(cidr("192.167.233.10/28"), ip("192.167.233.6"))"""))
+        assertEquals(1f, eval("""contains(cidr("192.167.233.11/32"), ip("192.167.233.11"))"""))
```

Comment:
Let's have a couple of tests using cidr at the beginning of the range? E.g., does `0.0.0.0/0` work and include all IPv4 addresses?

Reply:
Yes, missed this one, will add the test 😊